### PR TITLE
Adapted to Swift 3.

### PR DESCRIPTION
**Pull request by pilotniq**
I converted this project to Swift 3. 

I needed to do some conversions by hand, mainly for string indexes.

### CODE REVIEW

File: HostingApp/ViewController.swift [hunk 1]
```diff
@@ -52,10 +52,16 @@ class HostingAppViewController: UIViewController {
         self.stats?.text = "(Waiting for keyboard...)"
     }
     
-    func keyboardDidChangeFrame(notification: NSNotification) {
-        let frameEnd: CGRect! = notification.userInfo?[UIKeyboardFrameEndUserInfoKey]?.CGRectValue
-        
-        if frameEnd.height == referenceHeight {
+    func keyboardDidChangeFrame(_ notification: Notification) {
+      
+//      let frameEnd: CGRect = (notification.userInfo?[UIKeyboardFrameEndUserInfoKey] as CGRectValue).cg
```
Comment (simoncorstonoliver) [inline HostingApp/ViewController.swift]: Remove commented code or uncomment.

File: HostingApp/ViewController.swift [hunk 2]
```diff
@@ -52,10 +52,16 @@ class HostingAppViewController: UIViewController {
...
+//        let frameEnd: CGRect! = (notification as NSNotification).userInfo?[UIKeyboardFrameEndUserInfoKey]?.cgRectValue
```
Comment (simoncorstonoliver) [inline HostingApp/ViewController.swift]: Remove commented code or uncomment.

File: Keyboard/Catboard.swift [hunk 3]
```diff
@@ -17,26 +17,27 @@ let kCatTypeEnabled = "kCatTypeEnabled"
...
         
-        if !NSUserDefaults.standardUserDefaults().boolForKey(kCatTypeEnabled) {
+        if !UserDefaults.standard.bool(forKey: kCatTypeEnabled) {
             InsertText(keyOutput)
             return
         }
         
-        if key.type == .Character || key.type == .SpecialCharacter {
+        if key.type == .character || key.type == .specialCharacter {
+            // context will be a String?
```
Comment (simoncorstonoliver) [inline Keyboard/Catboard.swift]: Superfluous comment.

File: Keyboard/Catboard.swift [hunk 4]
```diff
@@ -45,14 +46,15 @@ class Catboard: KeyboardViewController {
                 }
                 
                 var index = context!.endIndex
-                
-                index = index.predecessor()
+              
+              index = context!.index( before: index )
+                // index = context!.endIndex(before: index)
```
Comment (simoncorstonoliver) [inline Keyboard/Catboard.swift]: Uncomment or delete line.

File: Keyboard/CustomNavigationController.swift [hunk 5]
```diff
@@ -54,12 +64,14 @@ class CustomNavigationController : UINavigationController {
     convenience init (parent: KeyboardViewController)
     {
         self.init()
-        self.parent = parent
+      // erl 2016-08-29
```
Comment (simoncorstonoliver) [inline Keyboard/CustomNavigationController.swift]: Superfluous comment

File: Keyboard/CustomNavigationController.swift [hunk 6]
```diff
@@ -17,28 +17,38 @@ class CustomNavigationController : UINavigationController {
...
 
         if countViews == 3 { // The nav bar itself + dummy VC (to force the back button) + settings VC
 
             countViews = 0
-            super.popViewControllerAnimated(false)
-            self.dismissViewControllerAnimated(false, completion: nil) // Nav bar goes away, revealing keyboard again
-            self.parent?.ChangeKeyboardLanguage(CurrentLanguageCode()) // But no event triggers the keyboard redrawing e.g. to account for selecting a different layout; so explicitly redo the kbd
+            super.popViewController(animated: false)
+            self.dismiss(animated: false, completion: nil) // Nav bar goes away, revealing keyboard again
+          // changed parent to myParent
```
Comment (simoncorstonoliver) [inline Keyboard/CustomNavigationController.swift]: Superfluous

File: Keyboard/CustomNavigationController.swift [hunk 7]
```diff
@@ -54,12 +64,14 @@ class CustomNavigationController : UINavigationController {
     convenience init (parent: KeyboardViewController)
     {
         self.init()
-        self.parent = parent
+      // erl 2016-08-29
+        parent.addChildViewController( self )
+//        self.parent = parent
```
Comment (simoncorstonoliver) [inline Keyboard/CustomNavigationController.swift]: Uncomment or remove.

**Comment by simoncorstonoliver** [comment]
Looks good thanks. Just a few places with commented out code -- either remove if not needed or uncomment if it was commented out for debugging.

**Comment by pilotniq** [comment]
Comments removed.

There's one thing I wasn't sure how to handle - have a look at Keyboard/CustomNavigationController.swift, around line 23, where I got an error because I couldn't override a stored var for 'parent'. I made a new variable with a getter called 'myParent' of the appropriate type, but you might want to handle it otherwise.

**Comment by simoncorstonoliver** [comment]
Pulled and built locally. Long-press on the globe then slide across to the gears icon causes an exception that kills the keyboard.

The local var name "parent" was poorly chosen (by me). There's already a "parent" for the UIController.

What I was trying to capture with the local "parent" is that I need a handle to the original keyboard so I can tell it to reload itself when we quit from the options.

So I think what we need is a simple local var:

```
var parentKeyboardVC: KeyboardViewController? = nil
```

then we have to assign to it when the convenience init gets called

self.parentKeyboardVC = parent

and use it to ensure we ChangeKeyboardLanguage

self.parentKeyboardVC?.ChangeKeyboardLanguage(CurrentLanguageCode()) 

However, still getting crashes so let me tinker with it a bit.

**Pull request merged**